### PR TITLE
VerifiedROBlobs in initial-sync

### PR DESCRIPTION
**Pull request by kasey**
# Correctly verify VerifiedROBlobs in Initial-sync 
Initial-sync is currently using the fake blob verification path, which we established as a temporary tool to help refactor across multiple pull requests. This PR changes initial-sync to the verification stack used in gossip processing, with some customizations to account for the batched nature of initial-sync.

## AvailabilityStore
This PR resurrects an earlier design that we implemented for blob caching called `AvailabilityStore`:

https://github.com/prysmaticlabs/prysm/blob/cd0679e24235fda9fc0d6652ea23733b4f1bf458/beacon-chain/das/iface.go#L12-L15

The idea of this pattern is that we can defer verification and persistence of `ROBlobs` until we are ready to perform the DA check for them. When `Persist` is called, the blob(s) are added to a cache in the store. When `IsDataAvailable` is called for the corresponding block, the blobs are verified and saved to disk, and a non-error nil value is returned if all of these steps were successful and data availability is guaranteed. This pattern will allow us to implement an alternate `AvailabilityStore` implementation for gossip processing to encapsulate the da check in a way that works for gossip and rpc sync. This matters because we want to use the same block processing code for both in the case of syncing non-finalized blocks via rpc.

One awkward side effect of this change is that until we introduce the gossip processing version, we need to pass a nil value in all the calls to ReceiveBlock that don't use the AvailabilityStore: https://github.com/prysmaticlabs/prysm/blob/cd0679e24235fda9fc0d6652ea23733b4f1bf458/beacon-chain/sync/subscriber_beacon_blocks.go#L30

onBlock internally checks for a nil value here and uses the current gossip version of the da check if the `AvailabilityStore` argument is nil: https://github.com/prysmaticlabs/prysm/blob/cd0679e24235fda9fc0d6652ea23733b4f1bf458/beacon-chain/blockchain/receive_block.go#L115-L123

### AvailabilityStore scope
Special note to keep in mind if reviewing AvailabilityStore in isolation is that it's scope is limited to a single batch.

## How batch verification is different
Gossip verification uses forkchoice to verify that a parent has been seen, is valid, to obtain the slot of the parent, and to determine whether it descends from finalized. Batch processing throws a wrench into this design, because forkchoice is only updated at the end of a batch of blocks.

There are also some optimizations that are easy when we verify all sidecars for a block together, like batching the kzg verification. 

### das.BlobBatchVerifier
https://github.com/prysmaticlabs/prysm/blob/cd0679e24235fda9fc0d6652ea23733b4f1bf458/beacon-chain/das/availability.go#L25-L32
Unlike the base `verification.BlobVerifier`, which exposes individual methods for each kind of verification, AvailabilityStore wants to just check them all together, leaving the order of checks up to the package using it.

## changes in processFetchedDataRegSync

Previously this function would bend over backwards trying to process all blocks in the batch, even if a parent was missing. But this can only happen 1 of 2 ways: either the first block's parent is missing, or there are gaps in the sequence of blocks. In the first case, we won't be able to process anything after the first block, since they all build on it. In the second, gossip rules state that:
```Clients MUST respond with blocks that are consistent from a single chain within the context of the request. This applies to any step value. In particular when step == 1, each parent_root MUST match the hash_tree_root of the preceding block.```
So if there is a missing parent anywhere in the batch, we can assume the rest of it is useless because the peer is misbehaving.

`processFetchedDataRegSync` now uses the `validUnprocessed` method to filter out any previously processed blocks, closing the loop on the other edge case that previously led to the decision to keep looping over the batch if there was an error. So now, the function ensures that it is starting with an unseen block, and then exits as soon as there is an issue, which simplifies the whole thing.

## BlockWithVerifiedBlobs -> BlockWithROBlobs

Previously `BlockWithVerifiedBlobs` referred to a block where the blobs had undergone some very minimal checks. To avoid confusion with `VerifiedROBlob`, this type has been renamed `BlockWithROBlobs`.

## initialsync.BlobBatchVerifier
This type implements `das.BlobBatchVerifier` for initial-sync. This is where we handle the special nature of batch verification, ~~keeping an internal cache of seen parents to override the forkchoice-based checks where appropriate~~ relying on the DA check only being called after the corresponding block has been verified, and batch verifying kzg commitments. 

## BlobVerifier.SatisfyRequirement
This has been added to allow BlobBatchVerifier to go around the verifier assertion methods and directly satisfy requirements itself. This enables kzg batch verification and the workaround for forkchoice-based checks in batches.

## variadic kzg.Verify
Extends the kzg commitment verification code to work for one or many BlobSidecars.

## SlotNotTooEarly -> NotFromFutureSlot
I got some feedback from @jtraglia that the naming and code of this method was convoluted, so I took his naming suggestion and reworked the code to hopefully be more intuitive.

## FakeVerifyForTest / FakeVerifySliceForTest

These were added to replace `BlobSidecarNoop` and `BlobSidecarSliceNoop`. The naming hopefully is even more clear that these are working around strong verification only for tests. They take and use a `*testing.T` parameter to make them incredibly difficult to use outside of tests.

## verification interface and mocks
Usually I try to push interfaces to packages that consume a type, rather than the producer, but this seems to be one of the edge cases where it's appropriate to break that rule. In this case we have a set of methods that need to be satisfied together, and need to be mocked across multiple packages. So it is practical to share that interface and mock amongst those multiple places (init-sync, sync, and later backfill).

## Updated: Signature check
The BlobBatchVerifier now relies on the caller to have verified the signature of the ROBlock given to IsDataAvailable. The verification code only ensures that the root of the unsigned block matches the root of the blob's unsigned block header, and checks that the signature bytes match across the two values.

## TODO
I haven't added a lot of new test coverage in this PR. Please feel free to suggest areas that stick out particularly in need of testing. I will follow up with more commits to add test coverage in particular on the batch verifier.

### CODE REVIEW

File: beacon-chain/blockchain/kzg/validation.go [hunk 1]
```diff
@@ -32,9 +32,26 @@ func IsDataAvailable(commitments [][]byte, sidecars []*ethpb.DeprecatedBlobSidec
...
+	}
+	blobs := make([]GoKZG.Blob, len(sidecars))
+	cmts := make([]GoKZG.KZGCommitment, len(sidecars))
+	proofs := make([]GoKZG.KZGProof, len(sidecars))
+	for i, sidecar := range sidecars {
+		blobs[i] = bytesToBlob(sidecar.Blob)
+		cmts[i] = bytesToCommitment(sidecar.KzgCommitment)
+		proofs[i] = bytesToKZGProof(sidecar.KzgProof)
+	}
+	return kzgContext.VerifyBlobKZGProofBatch(blobs, cmts, proofs)
```
Comment (terencechain) [inline beacon-chain/blockchain/kzg/validation.go:54]: we'll probably wanna add a unit test for these

File: beacon-chain/das/iface.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+// AvailabilityStore describes a component that can verify and save sidecars for a given block, and confirm previously
+// verified and saved sidecars.
+// Persist guarantees that the sidecar will be available to perform a DA check
+// for the life of the beacon node process.
+// IsDataAvailable guarantees that all blobs committed to in the block have been
+// durably persisted before returning a non-error value.
+type AvailabilityStore interface {
+	IsDataAvailable(ctx context.Context, current primitives.Slot, b blocks.ROBlock) error
+	Persist(current primitives.Slot, sc ...blocks.ROBlob) error
```
Comment (terencechain) [inline beacon-chain/das/iface.go:18]: What do you think about passing in the current slot versus assuming the implementation that satisfies these interfaces can get its current slot? Like`s.CurrentSlot()`. I'm not suggesting a change but rather curious

Comment (kasey) [reply beacon-chain/das/iface.go:18]: This was a decision in the older branch so I'm trying to remember my reasoning - I'm pretty sure it was because I wanted to avoid infecting this code with the whole `WaitForClock` mess that would come with giving it a clock. When we get that initialization path refactored to remove the need for clock waiting I think we could easily change this to keep an internal `CurrentSlotter`.

File: beacon-chain/sync/initial-sync/verification.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	if err := bv.NotFromFutureSlot(); err != nil {
+		return vb, err
+	}
+	if err := bv.SlotAboveFinalized(); err != nil {
+		return vb, err
+	}
+	if err := bv.ValidProposerSignature(ctx); err != nil {
+		return vb, err
+	}
```
Comment (jtraglia) [inline beacon-chain/sync/initial-sync/verification.go:76]: I'm a bit concerned that the order of these checks isn't optimal.

File: beacon-chain/blockchain/kzg/validation.go [hunk 4]
```diff
@@ -32,9 +32,26 @@ func IsDataAvailable(commitments [][]byte, sidecars []*ethpb.DeprecatedBlobSidec
...
+func Verify(sidecars ...blocks.ROBlob) error {
+	if len(sidecars) == 0 {
+		return nil
+	}
+	if len(sidecars) == 1 {
+		return kzgContext.VerifyBlobKZGProof(
+			bytesToBlob(sidecars[0].Blob),
+			bytesToCommitment(sidecars[0].KzgCommitment),
+			bytesToKZGProof(sidecars[0].KzgProof))
+	}
```
Comment (jtraglia) [inline beacon-chain/blockchain/kzg/validation.go:45]: Having a case for single kzg verification might be unnecessary. The performance difference should be negligible.

**Review by jtraglia — COMMENTED** [review]
Looking good so far. I'm still reviewing, but here are some initial comments.

File: beacon-chain/blockchain/kzg/validation.go (hunk 4 above)
Comment (kasey) [reply beacon-chain/blockchain/kzg/validation.go:45]: This is just a convenience so that gossip verification can use the same method, not an attempt to optimize single-blob blocks.

Comment (jtraglia) [reply beacon-chain/blockchain/kzg/validation.go:45]: I don't quite understand. Why do you have a case for single-blob then? It would work without that, no?

Comment (kasey) [reply beacon-chain/blockchain/kzg/validation.go:45]: That's a fair point, I can change that since it makes the function simpler.

Comment (jtraglia) [reply beacon-chain/blockchain/kzg/validation.go:45]: We discussed this offline. I'm happy with the current state of things.

File: beacon-chain/sync/initial-sync/verification.go (hunk 3 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/verification.go]: should be resolved here: https://github.com/prysmaticlabs/prysm/pull/13351/commits/9ade4cf800439d243945fb7ab79fbb06264ff104

File: beacon-chain/das/cache.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+// organized as a map from BlobSidecar.Index->BlobSidecar.KzgCommitment.
+// This representation is convenient for comparison to a block's commitments.
+type dbidx [fieldparams.MaxBlobsPerBlock]bool
+
+// missing compares the set of BlobSidecars observed in the backing store to the set of commitments
+// observed in a block - cmts is the BlobKzgCommitments field from a block.
+func (idx dbidx) missing(expected int) []uint64 {
+	m := make([]uint64, 0, expected)
+	for i := 0; i < expected; i++ {
+		if !idx[i] {
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/das/cache.go:79]: Added 2 layers of protection here - the function will "truncate" the check if it's too large, and `IsDataAvailable` will error right up top if the block has too many commitments: https://github.com/prysmaticlabs/prysm/pull/13351/commits/17ec410167a43b91bc6aed5b1cee52ab6bcf8c1f

**Comment by kasey** [comment]
This PR need some changes to decouple from forkchoice and inherit signature verification from block processing, so I am converting it to draft while I'm away on vacation.

**Review by rauljordan — APPROVED** [review]
Love the simplicity of some of the design patterns being used here, especially with the cache ensure approach. Don't have anything to add here, but did go through every file

**Review by jtraglia — COMMENTED** [review]
I did a quick review of the new changes. Just noticed two small comment nits.

And yes, it does appear to be much simpler now. I like it 😄

File: beacon-chain/das/availability.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	// ignore their response and decrease their peer score.
+	sidecars, err := entry.filter(root, blockCommitments)
+	if err != nil {
+		return errors.Wrap(err, "incomplete BlobSidecar batch")
+	}
+	// Do thorough verifications of each BlobSidecar for the block.
+	// Same as above, we don't save BlobSidecars if there are any problems with the batch.
+	vscs, err := s.verifier.VerifiedROBlobs(ctx, b, sidecars)
+	if err != nil {
+		me, ok := err.(verification.VerificationMultiError)
```
Comment (terencechain) [inline beacon-chain/das/availability.go:108]: This wont work on wrapped error. Do you want to use `errror.as`?
```go
		var me verification.VerificationMultiError
		ok := errors.As(err, &me)
```

File: beacon-chain/das/cache.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+// cacheKey includes the slot so that we can easily iterate through the cache and compare
+// slots for eviction purposes. Whether the input is the block or the sidecar, we always have
+// the root+slot when interacting with the cache, so it isn't an inconvenience to use both.
+type cacheKey struct {
+	slot primitives.Slot
+	root [32]byte
+}
+
+type cache struct {
+	entries map[cacheKey]*cacheEntry
```
Comment (terencechain) [inline beacon-chain/das/cache.go:28]: Given this cache is unbounded, have you considered adding metrics? I'm worried there would be dangling items around. Can be a follow up PR

File: beacon-chain/sync/initial-sync/verification.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		if blk.Root() != scs[i].BlockRoot() {
+			return nil, ErrBatchBlockRootMismatch
+		}
+	}
+	// Verify commitments for all blobs at once. verifyOneBlob assumes it is only called once this check succeeds.
+	if err := batch.verifyKzg(scs...); err != nil {
+		return nil, err
+	}
+	vs := make([]blocks.VerifiedROBlob, len(scs))
+	for i := range scs {
```
Comment (terencechain) [inline beacon-chain/sync/initial-sync/verification.go:73]: Why not just use the earlier `for i := range scs {` for this. I think you can just do `batch.verifyKzg(scs...)` first

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 9]
```diff
@@ -159,47 +159,46 @@ func (s *Service) processFetchedDataRegSync(
...
-			if err != nil {
-				log.WithField("root", b.Block.Root()).WithError(err).Error("blobs failed verification")
-				continue
-			}
-			for i := range verified {
-				if err := s.cfg.BlobStorage.Save(verified[i]); err != nil {
-					log.WithError(err).Warn("Failed to save blob sidecar")
-				}
-			}
+		if err := avs.Persist(s.clock.CurrentSlot(), b.Blobs...); err != nil {
```
Comment (terencechain) [inline beacon-chain/sync/initial-sync/round_robin.go:177]: There's a potential DA memory exploit here if we persist without ensuring the block is valid. Here's a summary:
- First, we save blobs to memory in the sync package.
- Next, in the blockchain package, we verify the blob's validity by retrieving it from the cache. We then check if the blob and block are aligned, including kzg checks. Regardless of whether they pass or fail, we always delete the blob from the cache.
- We verify the blob after verifying the block.
- It's important to note that someone could provide blobs referencing a bad block. In this case, the blobs would remain stuck in memory.

To fix this, we have a few options:
- Implement `defer s.cache.delete(key)` at the `ReceiveBlock` level to ensure the blob is always cleared when the function ends
- Add a TTL to the blob cache
- Avoid persisting blobs to the cache unless the block is confirmed valid. This change might be less elegant, and I'm not entirely convinced it's the best approach

File: beacon-chain/das/cache.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
+package das
+
+import (
+	"bytes"
+
+	errors "github.com/pkg/errors"
```
Comment (terencechain) [inline beacon-chain/das/cache.go:6]: do we need to alias here?

**Review by terencechain — CHANGES_REQUESTED** [review]
Have a concern about memory exploit

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 9 above)
Comment (terencechain) [reply beacon-chain/sync/initial-sync/round_robin.go:177]: Correction: The scope of AVS is per batch. This is not an issue

File: beacon-chain/das/cache.go (hunk 7 above)
Comment (terencechain) [reply beacon-chain/das/cache.go:28]: No need. The scope of AVS is per batch

File: beacon-chain/das/availability.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+		return errors.Wrapf(err, "could check data availability for block %#x", b.Root())
+	}
+	// Return early for blocks that are pre-deneb or which do not have any commitments.
+	if blockCommitments.count() == 0 {
+		return nil
+	}
+
+	key := keyFromBlock(b)
+	entry := s.cache.ensure(key)
+	defer s.cache.delete(key)
```
Comment (terencechain) [inline beacon-chain/das/availability.go:95]: You probably don't need to delete here. if the cache's life time is per batch level. Just let it expire

**Review by terencechain — DISMISSED** [review]
I've left some optional feedback and corrected my prev missunderstanding

File: beacon-chain/das/cache.go (hunk 7 above)
Comment (kasey) [reply beacon-chain/das/cache.go:28]: > Have a concern about memory exploit
In reference to the above, which is resolved/clarified now, right?

File: beacon-chain/das/cache.go (hunk 10 above)
Comment (kasey) [reply beacon-chain/das/cache.go:6]: huh not sure why ide added the import that way, will tidy it up.

File: beacon-chain/sync/initial-sync/verification.go (hunk 8 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/verification.go:73]: I want to do the cheap signature check before the expensive kzg verification.

File: beacon-chain/verification/blob.go [hunk 12]
```diff
@@ -44,12 +44,31 @@ var GossipSidecarRequirements = []Requirement{
 	RequireSidecarProposerExpected,
 }
 
+// InitsyncSidecarRequirements is the list of verification.Requirements to be used by the init-sync service
+// for batch-mode syncing. Because we only perform batch verification as part of the IsDataAvailable method
+// for blobs after the block has been verified, and the blobs to be verified are keyed in the cache by the
+// block root, it is safe to skip the following verifications.
+// RequireValidProposerSignature
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/verification/blob.go:51]: It should be removed from this list - I mistakenly added it to the list of skipped checks because I briefly forgot that the block root doesn't integrate the signature, so the signature still needs to be explicitly checked.

File: beacon-chain/das/availability.go (hunk 6 above)
Comment (kasey) [reply beacon-chain/das/availability.go:108]: `errors.As` should recursively `Unwrap` errors until it finds one that matches the given type?

Comment (kasey) [reply beacon-chain/das/availability.go:108]: oh woops I misunderstood the context, yeah it would be safer to use errors.As in case this gets wrapped in the future :+1:

File: beacon-chain/das/availability.go (hunk 11 above)
Comment (kasey) [reply beacon-chain/das/availability.go:95]: i think i'm going to leave it, might as well try to free the memory as soon as possible

**Review by terencechain — APPROVED** [review]

**Review by terencechain — APPROVED** [review]
